### ISSUE TITLE
experimentalSourceRewriting performance issue

### ISSUE THREAD

**Issue by laurentenhoor** [cypress-io/cypress#7565]
### Current behavior:

```json
{
	"experimentalSourceRewriting": true
}
```
After enabling `experimentalSourceRewriting`, a single `VISIT` takes about 5-10 times as long. An unacceptable increase for our pipelines. HTTP requests to `GET` the page's resources seems to be significantly slower.

### Desired behavior:

We want the HTTP requests to be as performant as without `experimentalSourceRewriting`.

### Test code to reproduce
I created this very simple PR: https://github.com/cypress-io/cypress-test-tiny/pull/59
Please try to disable/re-enable `experimentalSourceRewriting`.
Have a look at the difference in time that the following test takes:
```js
describe('page', () => {
  it('works', () => {
    cy.visit('https://example.cypress.io')
  })
})
```
if `experimentalSourceRewriting: true` it takes ~1583ms
if `experimentalSourceRewriting: false` it takes ~300ms

### Versions

 Cypress 4.6.0

**Comment by jennifer-shehane** [comment]
There does appear to be ~1000ms increase in the test time for visiting this specific URL when `experimentalSourceRewriting: true` is set.

`spec.js`

```js
Cypress.on('test:after:run', (attributes) => {
  console.log('Test "%s" has finished in %dms', attributes.title, attributes.duration)
})

describe("Testing", () => {
  Array(15).fill(0).forEach((val, i) => {
    it(`test ${i}`, () => {
      cy.visit('https://example.cypress.io')
    })
  })
})
```

### With `experimentalSourceRewriting: true` 

<img width="535" alt="Screen Shot 2020-06-02 at 3 55 39 PM" src="https://user-images.githubusercontent.com/1271364/83503909-d371b080-a4e9-11ea-9cd3-48e0fce016c1.png">


### With `experimentalSourceRewriting: false`

<img width="537" alt="Screen Shot 2020-06-02 at 3 55 59 PM" src="https://user-images.githubusercontent.com/1271364/83503957-e5535380-a4e9-11ea-8aa1-c84e0e64a15c.png">

**Comment by a8trejo** [comment]
I am experiencing a similar Issue, a little worse, 
Context, 
- For my Test Suite I'm using Cucumber features
- There is a specific test with a login which only works when I set "experimentalSourceRewriting" as true
- When I set "experimentalSourceRewriting" as true, in all my other Scenarios (same feature file) other cy.visit never load the page, I already tried increasing the pageLoadTimeOut up to ridiculous 80 seconds and it just never loads.

I am trying to use `Cypress.config('experimentalSourceRewriting',true)` only for that specific test where I need it, and the value supposedly changes (if I perform a cy.log it comes as true) but then I will get the 404 error as if it was not configured.

Is there a way to set Single test configuration for a specific cucumber Scenario? (Setting 'experimentalSourceRewriting',true in this case)

**Comment by a8trejo** [comment]
In case it helps anyone, the workaround I found for Cucumber Tests was Tagging that specific Scenario with something like @experimental and then use the command line to run only that Scenario with that configuration:

`npx cypress run --spec **/Name.feature -e TAGS='@experimental' --config experimentalSourceRewriting=true`

And for all my other tests (cypress.json has `"experimentalSourceRewriting" : false` ):
`npx cypress run -e TAGS='not @experimental'`

For normal Cypress mocha tests I guess they could pass the single test Configuration like this? (Haven't tried it)
https://docs.cypress.io/es/guides/references/configuration.html#Single-test-configuration

My workaound however gives me a new Issue as the cucumberJSON report generation only creates one report per feature file, and as far as I know still has no timestamp option, hence I only get one report out of my 2 executions of that feature and I need cucumberJSON format for my CI/CD cloud integration with JiraXray

**Comment by modern-sapien** [comment]
[**Issue 8128**](https://github.com/cypress-io/cypress/issues/8128) is similar and has a workaround.

**Comment by JoostK** [comment]
Coming from #29438, I dug into it with a profiler and found my case to be primarily due to cloning the AST in `recast`, where converting some linear searches into binary searches reduces the time to 5s (from 100s+, although I don't have exact numbers at the moment because I profiled using a different laptop and didn't capture the original duration on this machine)

Even with that addressed—I want have to look into it further to better understand why this makes such a difference here—there's more room for improvement, where reusing the original JS instead of printing a potentially unmodified AST is super wasteful. Potential saving of ~1.8s, bringing the total down to 3.2s for the file I was seeing 100s+ rewrite times for. Still not great and room for improvement, but getting somewhere.

**Update**: after some more optimizations, we're down to ~1.3s. Mostly in an experimental state still.

**Comment by jennifer-shehane** [comment]
@JoostK It'd be helpful, if possible, to get those optimizations back in as a PR to update this logic.

**Comment by JoostK** [comment]
@jennifer-shehane yeah definitely, I will have to carve out some more time for it though. Hopefully somewhere this month.

**Comment by JoostK** [comment]
Okay, I have two PRs out:

- https://github.com/benjamn/recast/pull/1399
  - This offers the biggest improvement for the issue I was seeing, going from ~120s to 1.2s.
- https://github.com/cypress-io/cypress/pull/29540
  - This implements some optimizations in Cypress' use of recast to allow my case to go from 4.4s to 2.3s (provided that the recast fix has been patched in).

**Comment by jennifer-shehane** [comment]
Thanks for the detailed reproduction here, and a big thank you to @JoostK for profiling this and driving the fixes upstream and in Cypress. 🙏

The performance of experimentalSourceRewriting has been substantially improved through two changes:
- benjamn/recast#1399: replaced linear searches with binary searches inside recast itself. This was the biggest win, taking a large bundle from ~120s down to ~1.2s.
- #29540 - optimized Cypress’s own use of recast by (1) skipping the re-print step entirely when the AST visitor made no changes, and (2) deferring the expensive scope lookup so it only runs for identifiers that might actually need rewriting.

Together these brought a 2.6MB bundle that requires no rewriting from ~4.4s down to ~2.3s, on top of the recast improvement above.

The Cypress changes were released in v13.13.0, and the recast fix is included via the version Cypress now ships (recast@0.23.11).

One note for anyone landing here: source rewriting is fundamentally a parse → transform → re-print pipeline, so it will always carry some overhead compared to the default rewriter, but it should no longer be the 5–10× regression originally reported.

Since the underlying performance work has been merged and released, we’re going to close this issue. 

If you’re still seeing unacceptable rewriting times on a recent version of Cypress, please open a new issue with a reproduction (ideally with a profile or the bundle sizes involved) and we’ll take another look.

**Comment by JoostK** [comment]
Hello AI!

> via the version Cypress now ships (recast@0.23.11).

https://github.com/benjamn/recast/pull/1399 unfortunately hasn't been merged, so this point is moot.
